runtime/src/bank.rs

```diff
@@ -675,14 +686,82 @@ impl Bank {
 
     /// Return the checkpointed bank that should be used to generate a leader schedule.
     /// Return None if a sufficiently old bank checkpoint doesn't exist.
-    pub fn leader_schedule_bank(&self) -> Option<Arc<Bank>> {
-        let epoch_slot_height = self.slot_height() - self.slot_index();
+    fn leader_schedule_bank(&self, epoch_height: u64) -> Option<Arc<Bank>> {
+        let epoch_slot_height = epoch_height * self.slots_per_epoch();
         let expected = epoch_slot_height.saturating_sub(self.leader_schedule_slot_offset);
         self.parents()
             .into_iter()
             .find(|bank| bank.slot_height() <= expected)
```

Comment:
The slot_leader() function may be be called after the bank at `epoch_slot_height.saturating_sub(self.leader_schedule_slot_offset);` is merged into the root, in which case this bank for generating for leader schedule will no longer be found. Will this be addressed in a different PR where we do memoization?

Reply:
Yes, separate PR.  I see 3 options: memoize at each epoch; memoize only in the root; reject some attempts to merge as "too soon".  The first option seems most likely.

Reply:
@carllin, I updated the code so that we'd only need to memoize `staked_nodes()`, not the leader schedule. Figured that might have utility elsewhere (@sagar-solana, gossip?), and it let me reduce the LeaderSchedule to almost no code.